### DISCUSSION TITLE
Point of useTransition for server actions?

### DISCUSSION THREAD

**Discussion by codinginflow** [vercel/next.js#51121] [App Router]
I don't understand what the point of this `useTransition` wrapper is in the Next.js server action docs:

https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-using-starttransition

I know about `useTransition` but I don't get why it's used here. I even tried it out and didn't notice a difference than just calling the server action in the click handler directly (like they do in the chapter below).

**Comment by devjiwonchoi** [comment]
**The main difference is whether you use `form` or not.**

> You can achieve this by using startTransition provided by the useTransition hook, which can be useful if you want to use Server Actions outside of forms, buttons, or inputs.

This is just another option to use Server Actions, in which for `forms`, you are recommended to use `action` and `formAction`, and for the other cases, you can use `useTransition`.

## Reference
- [Server Actions `startTransition`](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-using-starttransition)

**Reply by codinginflow** [reply]
I read this part of the docs but it doesn't explain **why** the `useTransition` is necessary. You can await a server action in a client component without `useTransition` and it won't make a difference.

**Reply by devjiwonchoi** [reply]
I get what you mean. The advantage of using `useTransition` is based on the `Server Mutation`. 
You can handle the transition state and pending state by `isPending`, and are able to squeeze in between the pending state and transition state and **mutate** the actions such as `revalidate`, `redirect`, etc.

By `awaiting` server actions, it is just executing the functions asynchronously, and you don't have the opportunity to manage the actions by default.

## Reference
- [Server Actions without `useTransition`](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-without-starttransition)
- [Server Mutations](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#server-mutations)

**Reply by codinginflow** [reply]
> I get what you mean. The advantage of using `useTransition` is based on the `Server Mutation`. You can handle the transition state and pending state by `isPending`, and are able to squeeze in between the pending state and transition state and **mutate** the actions such as `revalidate`, `redirect`, etc.
> 
> By `awaiting` server actions, it is just executing the functions asynchronously, and you don't have the opportunity to manage the actions by default.
> 
> ## Reference
> * [Server Actions without `useTransition`](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-without-starttransition)
> * [Server Mutations](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#server-mutations)

Sorry, but I don't understand it. What do you mean by "managing actions" or "squeezing in between the pending state and transition state"?

Can you give a pseudo-code example of when a server action would behave differently with and without `useTransition`?

**Reply by devjiwonchoi** [reply]
## With `useTransition`

While using `useTransition`, you can do **Server Mutation** such as `redirect`, and is able to catch the `isPending` state which is in between `onClick` and fully resolving the `actions`, and display the `Redirecting...` text.

```js
// With useTransition
'use client'

import { useTransition } from 'react'
import { redirectToGoogle } from '@/app/actions'

export function ServerActionButton() {
  let [isPending, startTransition] = useTransition()

  // Displays `Redirecting...` text until executing `redirectToGoogle()`
  if (isPending) {
    return <p>Redirecting...</p>
  }

  return (
    <button onClick={() => startTransition(() => redirectToGoogle())}>
      Redirect to Google
    </button>
  )
}
```

https://github.com/vercel/next.js/assets/120007119/1bf316dd-2065-497a-b20c-775b7fe1783b

## Without `useTransition`  - Trying to Server Mutate (Redirect)

Without `useTransition`, you cannot Mutate actions which means you cannot also redirect, and the previous example won't work.

```js
// page.tsx
import { ServerActionButton2 } from '@/components/ServerActionButton2'
import { redirectToGoogle } from '@/app/actions'

export default function Home() {
  return <ServerActionButton2 redirectToGoogle={redirectToGoogle} />
}
```

```js
// Without useTransition
'use client'

export function ServerActionButton2({
  redirectToGoogle,
}: {
  redirectToGoogle: () => Promise<void>
}) {
  return (
    <button onClick={async () => await redirectToGoogle()}>
      Redirect to Google
    </button>
  )
}
```

https://github.com/vercel/next.js/assets/120007119/156e7305-1b10-4250-a486-5f7c07a9f4aa

## Without `useTransition`  - Does not Server Mutate

You can only use actions that do not Mutate the actions.

```js
// page.tsx
import { ServerActionButton2 } from '@/components/ServerActionButton2'
import { sayHello } from '@/app/actions'

export default function Home() {
  return <ServerActionButton2 sayHello={sayHello} />
}
```

```js
'use client'

export function ServerActionButton2({
  sayHello,
}: {
  sayHello: () => Promise<void>
}) {
  return (
    <button onClick={async () => await sayHello()}>
      Say Hello to Coding in Flow
    </button>
  )
}
```

```js
// action.ts
'use server'

export async function sayHello() {
  console.log('Hello Coding in Flow!')
}
```

![Screenshot 2023-06-11 at 9 29 54 PM](https://github.com/vercel/next.js/assets/120007119/1a926cf1-5f7f-4363-8a66-af1ed393ef48)

Since it is a React feature, it is described well [here](https://react.dev/reference/react/useTransition).

## Conclusion
The main advantage of `useTransition` is that you can catch the `isPending` state and `Server Mutation`.

## Reference
- [React `useTransition`](https://react.dev/reference/react/useTransition)
- [Server Actions with `useTransition`](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-using-starttransition)
- [Server Actions without `useTransition`](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-without-starttransition)
- [Server Mutations](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#server-mutations)

**Reply by codinginflow** [reply]
What is the implementation of your server action here? Why would the server action care what the client does? This doesn't look right to me.

**Edit:** Ok, I tried it out myself and redirect indeed doesn't work. But why? And why do the docs include `revalidatePath` into this which doesn't have the same problem?

**Reply by devjiwonchoi** [reply]
Note that `revalidatePath` won't work without `useTransition` since it is part of Server Mutation, and the example in the docs about `revalidatePath` is within [Custom invocation using startTransition](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-using-starttransition).

**Reply by codinginflow** [reply]
> Note that `revalidatePath` won't work without `useTransition`

It does work. I've been using it without transition. That's why I am on this journey trying to find out why the docs tell me to use `useTransition`.

**Reply by devjiwonchoi** [reply]
Yeah, I've tested it out myself and... why does it work? Gotta dig more into it.

**Reply by devjiwonchoi** [reply]
@codinginflow I'm assuming that the Revalidating itself is supported by default in Server Actions, therefore it is not restricted by the usage of `form` or `useTransition`.

> Server Actions can be used to revalidate data on-demand by path ([revalidatePath](https://nextjs.org/docs/app/api-reference/functions/revalidatePath)) or by cache tag ([revalidateTag](https://nextjs.org/docs/app/api-reference/functions/revalidateTag)).

**Reply by codinginflow** [reply]
> @codinginflow I'm assuming that the Revalidating itself is supported by default in Server Actions, therefore it is not restricted by the usage of `form` or `useTransition`.
> 
> > Server Actions can be used to revalidate data on-demand by path ([revalidatePath](https://nextjs.org/docs/app/api-reference/functions/revalidatePath)) or by cache tag ([revalidateTag](https://nextjs.org/docs/app/api-reference/functions/revalidateTag)).

Thank you for the clarification and your help so far! Maybe we get more info about this in the future.
